### PR TITLE
Add bevy_dev_tools crate

### LINKED ISSUE #11309: Create bevy_debug crate
## What problem does this solve or what need does it fill?

With the start of the bevy_editor prototype, it has become an important discussion to address the issue of what is needed to move forward with this project.
Most people agree that to do this, we need several debugging tools so that the user can enjoy a complete editing system.
However, another issue in this regard deserves discussion: Should we create a crate just for these debug tools, so that they are all housed in a separate block?
The advantages of this type of approach are numerous: If someone does not need such tools, we can disable everything as easily as disable a feature, in addition, the nesting of these tools will help in understanding the code in the future
So, as advantages, this feature would provide:
- Nesting of debug tools for easy deactivation/activation
- Management of these tools by new contributors made easier
- Making it easier to visualize what we still need for the new editor

## What solution would you like?

The creation of the bevy_debug crate (or similar) for the purpose of storing those tools and the moving of all debugs tools already existent in the engine (this last part is not very necessary: We would probably have to define what is only a debug tool and what is not for this purpose) 

## What alternative(s) have you considered?

We can simply have a debug mod in all crates for the purpose of storing the debug tools of that said crate there
The problem with this approach would probably be debug tools that are of interest of multiple crates: Where to store this kind of tool?

## Additional context

[Useful discussion](https://discord.com/channels/691052431525675048/1192567920346271755/1192629117112307752) on this issue.

[The proposal](https://discord.com/channels/691052431525675048/856713292067766382/1194798366991847424) of the new editor prototyping.

### PR DESCRIPTION
**Pull request by matiqo15**
# Objective

- Resolves #11309

## Solution

- Add `bevy_dev_tools` crate as a default feature.
- Add `DevToolsPlugin` and add it to an app if the `bevy_dev_tools` feature is enabled.

`bevy_dev_tools` is reserved by @alice-i-cecile, should we wait until it gets transferred to cart before merging?

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Added the `A-Editor` label: I'm going to group the dev-tools and editor PRs and issues together for triage until that proves to be a bad idea.

I'm going to wait on @cart's sign-off before merging this, but I think we're starting to get a clearer vision of what this crate might look like and can unblock editor-adjacent work *now*.

File: crates/bevy_dev_tools/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:3]:
```suggestion
//! focused on interactive in-app debugging. 
```

File: crates/bevy_dev_tools/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:1]: This can be moved to the [lints] table in Cargo.toml now :)

File: crates/bevy_dev_tools/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:7]:
```suggestion
/// Enables interactive debugging tools in an [`App`].
```

**Review by alice-i-cecile — APPROVED** [review]
I think we can make the docs clearer, by focusing on a well-defined niche. That's non-blocking though, and I think this is the right step at this time.

File: crates/bevy_dev_tools/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:7]: We probably also want to add a note about how you should turn these off when shipping.

File: Cargo.toml
Comment (pablo-lua) [inline Cargo.toml:61]: Do we really want this to be enabled by default 🤔 ?   If we go with "disable this in the release build" approach, wouldn't that hinder disabling this feature?
We would have to do something like that in the TOML:
```toml
[dependencies]
bevy = {version = "0.12.1", default-features = false, features = ["all features", "except the debug feature"]}
[dev-dependencies]
bevy = {version = "0.12.1"}
```

File: crates/bevy_dev_tools/src/lib.rs
Comment (matiqo15) [reply crates/bevy_dev_tools/src/lib.rs:3]: Do we want this crate to be only for debugging tools? There *might* be some other tools, that would improve workflow but are not designed for debugging.

File: Cargo.toml
Comment (BD103) [reply Cargo.toml:61]: There are two different approaches that I could see being used.

### Disable on release mode

You could wrap `fn build()` in a `#[cfg(debug_assertions)]` so that the plugin only does things in debug mode.

```rust
#[cfg(debug_assertions)]
fn build(&self, app: &mut App) {
    ...
}
```

I don't think this is a good idea for a multitude of reasons. First, it's inconsistent with other developer crates Bevy provides. [Dynamic linking](https://bevyengine.org/learn/book/getting-started/setup/#enable-fast-compiles-optional) is opt-in, so I believe that devtools should be treated the same. Additionally, [the setup guide](https://bevyengine.org/learn/book/getting-started/setup/#compile-with-performance-optimizations) recommends turning on `opt-level = 3` for all dependencies, including Bevy. I'm not positive, but that may turn off debug assertions and thus devtools.

### Non-default feature

I think this is the better option. This only issue is that developers will need to know that the devtools exist in order to use them. This is fine in my opinion, as it will work similar to `bevy_dylib`. And since developers will have to manually enable it, they will know to disable it in release mode. (At least I hope.)

Comment (NthTensor) [reply Cargo.toml:61]: Non-default feature sounds best to me. Maybe `DevToolsPlugin` should emit a warning on build in production.

File: crates/bevy_dev_tools/src/lib.rs
Comment (matiqo15) [reply crates/bevy_dev_tools/src/lib.rs:1]: I believe this would need to be in the main Cargo.toml, and some crates are not fully documented. This is because we use lints from the workspace. We could also just repeat what has been set in the main Cargo.toml. 

I'd leave it as is because it's simpler and consistent with how it's done in other crates.

File: Cargo.toml
Comment (pablo-lua) [reply Cargo.toml:61]: > Emit a warning on production

A warning like "you are running the dev-tools on release, ignore if this is intentional"?

Comment (NthTensor) [reply Cargo.toml:61]: Exactly. I'm sure I will totally forget to disable it otherwise.

File: crates/bevy_dev_tools/src/lib.rs
Comment (BD103) [reply crates/bevy_dev_tools/src/lib.rs:1]: This can be configured per each crate. ([See announcement](https://blog.rust-lang.org/2023/11/16/Rust-1.74.0.html#lint-configuration-through-cargo)) If other crates don't use the `[lints]` table, then perhaps they need to be changed as well.

File: Cargo.toml
Comment (alice-i-cecile) [reply Cargo.toml:61]: Yeah, I think we should leave this in default features to make the examples and beginner experience run smoothly, but emit an `info!()` log about "you're running dev tools! This may give players undesired access to your game! Maybe turn it off when shipping?"

Leaving it off by default only helps stop the absolute beginners from shipping it, but experienced teams will do this too.

File: crates/bevy_dev_tools/src/lib.rs
Comment (alice-i-cecile) [reply crates/bevy_dev_tools/src/lib.rs:1]: Yep, we should migrate the other crates over too. Different PR though!

Comment (matiqo15) [reply crates/bevy_dev_tools/src/lib.rs:1]: Yes, but we are using workspace lints https://github.com/bevyengine/bevy/blob/8a523de8db28bfdc5b2a4c67c2022cc957075de6/Cargo.toml#L33-L41

And because of this, when we want to add lints only per crate we can't import lints from workspace because it results in: ```cannot override `workspace.lints` in `lints`, either remove the overrides or `lints.workspace = true` and manually specify the lints```

It's imported in  bevy_dev_tools/Cargo.toml via: 
```toml
[lints]
workspace = true
```

That's what I meant with:
> We could also just repeat what has been set in the main Cargo.toml.

File: Cargo.toml
Comment (pablo-lua) [reply Cargo.toml:61]: I can understand this point, offering the beginners the tools by default is good, but IMO the only problem is the pain of disabling it by listing all the default features (you will have to turn off the default-features, and them list all the default-features, except for the bevy_dev_tools). That wouldn't be a problem if cargo supported disabling one of default features with some sintax though. I do agree however that this "problem" may not be a problem, just a boilerplate in the Cargo.toml 
An example of the result, using all the default-features:
```toml
[dev-dependencies]
bevy = {version = "0.12.1"}

[dependencies]
bevy = {version = "0.12.1", features = [
  "animation",
  "bevy_asset",
  "bevy_audio",
  "bevy_gilrs",
  "bevy_scene",
  "bevy_winit",
  "bevy_core_pipeline",
  "bevy_pbr",
  "bevy_gltf",
  "bevy_render",
  "bevy_sprite",
  "bevy_text",
  "bevy_ui",
  "multi-threaded",
  "png",
  "hdr",
  "vorbis",
  "x11",
  "bevy_gizmos",
  "android_shared_stdcxx",
  "tonemapping_luts",
  "default_font",
  "webgl2",
  # all the actual default features, except for this one
]}
```
Well, this is a boilerplate, but nothing will go wrong for sure
One another option for beginners is strongly recomending the feature somewhere in the book, but that would probably not be as effective as that feature being a default-feature though

Comment (matiqo15) [reply Cargo.toml:61]: what about using `info!("dev tools are enabled")` in dev mode, and `warn!("you're running dev tools in release mode, this is not recommended")` in release mode - text used as an example

Comment (matiqo15) [reply Cargo.toml:61]: > I can understand this point, offering the beginners the tools by default is good, but IMO the only problem is the pain of disabling it by listing all the default features (you will have to turn off the default-features, and them list all the default-features, except for the bevy_dev_tools). That wouldn't be a problem if cargo supported disabling one of default features with some sintax though. I do agree however that this "problem" may not be a problem, just a boilerplate in the Cargo.toml 
> An example of the result, using all the default-features:
> ```toml
> [dev-dependencies]
> bevy = {version = "0.12.1"}
> 
> [dependencies]
> bevy = {version = "0.12.1", default-features = false, features = [
>   "animation",
>   "bevy_asset",
>   "bevy_audio",
>   "bevy_gilrs",
>   "bevy_scene",
>   "bevy_winit",
>   "bevy_core_pipeline",
>   "bevy_pbr",
>   "bevy_gltf",
>   "bevy_render",
>   "bevy_sprite",
>   "bevy_text",
>   "bevy_ui",
>   "multi-threaded",
>   "png",
>   "hdr",
>   "vorbis",
>   "x11",
>   "bevy_gizmos",
>   "android_shared_stdcxx",
>   "tonemapping_luts",
>   "default_font",
>   "webgl2",
>   # all the actual default features, except for this one
> ]}
> ```
> Well, this is a boilerplate, but nothing will go wrong for sure
> One another option for beginners is strongly recomending the feature somewhere in the book, but that would probably not be as effective as that feature being a default-feature though

You could also just disable DevToolsPlugin from DefaultPlugins, should work the same from what I know

Comment (mockersf) [reply Cargo.toml:61]: I also think this should not be part of the default features

**Comment by mockersf** [comment]
> should we wait until it gets transferred to cart before merging?

Not necessarily, but I think we should have something in those tools before merging. You could try moving https://github.com/bevyengine/bevy/blob/main/crates/bevy_app/src/ci_testing.rs

**Comment by alice-i-cecile** [comment]
I'm fine to move that into here as part of the initial PR. There's a number of other things I want to move and add, but this is a good uncontroversial first option.

File: Cargo.toml
Comment (rlidwka) [reply Cargo.toml:61]: > Maybe DevToolsPlugin should emit a warning on build in production.

What's wrong with having dev tools in release mode?

Comment (NthTensor) [reply Cargo.toml:61]: Dev tools will probably add bloat and slow things down: profiling, version-control/undo-systems, ect add a lot of extra stuff. Most users won't want to ship them, and release mode is the best indicator that "ok, this build is maybe going to ship". I'm not suggesting that they shouldn't work in release, just that they log a warning to help prevent accidents.

But I guess that's a bridge we can cross when we come to it. The crate is empty. If we need to take this off default at a later point we can.

Comment (rlidwka) [reply Cargo.toml:61]: If you want an option to *not* compile it in `release`, it should be behind a new feature flag. This is very similar to `debug-render-3d` in rapier, which is a debug feature behind a flag.

Keep in mind that people (like me) would most likely want to include it in `release` for better debuggability and modding support in production (unless there's good reason not to, e.g. pulling in tracing libraries).

Emitting a warning in production is just wrong. The code is already compiled in by that point, so what is that warning trying to achieve besides being annoying in false positive cases?

**Comment by matiqo15** [comment]
Probably, we also want an option to enable only some tools
### 1st solution
We remove `bevy_dev_tools` from default features and lock every tool under a new feature - this is a simple solution but could grow out of control over time.
### 2nd solution
We create a resource to store information about dev tools (see DiagnosticStore).  There are a few reasons why I think this is better:

- Easier for users to add their own tools
- We could mark it as an editor/non-editor tool
- Show every enabled tool in the editor
- Possibly in the future, add/remove tool at runtime

I believe this could be done in a follow-up PR, to make it easier for review.

**Comment by alice-i-cecile** [comment]
Yep, IMO we add a few tools first, then start thinking about how to abstract them. I'd prefer to have it runtime configured too.

File: Cargo.toml
Comment (cart) [reply Cargo.toml:61]: This should not be a default crate for a number of reasons:
1. We currently operate under the assumption that `cargo build --release` should be a "shippable" binary. Practically, this is how people will use Bevy. Removing the default features and re-adding what you need is a nasty, constantly shifting, and arcane workflow in my opinion. This isn't something we can thrust on users by default. We can reconsider once we have an editor capable of facilitating these complexities, but even then I am dubious.
2. This is a "grab bag" crate in its current form (very loose bounds on scope, size and contents will likely balloon indefinitely), with little to no intended value for released apps. Something like that shouldn't be shipped by default imo.

I think this should be an opt-in cargo. Any examples that rely on the dev tools should add the feature as a requirement. For things that _should_ be enabled by default, they can just have their own owns with a clearer scope.

**Comment by cart** [comment]
Just left some thoughts in this thread: https://github.com/bevyengine/bevy/pull/11341/files#r1459692523

File: crates/bevy_dev_tools/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:25]:
```suggestion
///  creating a build into for release to the public.
```

File: crates/bevy_dev_tools/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:14]: There's a third, better way to do this.

Create your own crate feature e.g. `dev_mode`, which enables the `bevy_dev_tools` dependency (and any other dev tools you might have).

File: crates/bevy_dev_tools/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/lib.rs:12]:
```suggestion
/// Warning: It is not recommended to enable this in final shipped games or applications.
/// Dev tools provide a high level of access to the internals of your application,
/// and may interfere with ordinary use and gameplay.
```

**Review by doonv — APPROVED** [review]

**Comment by BD103** [comment]
#12305 is a feature that fits the best in `bevy_dev_tools`. What needs to be done for this PR to be merged, to unblock the simplified backtraces one?

File: crates/bevy_dev_tools/Cargo.toml
Comment (BD103) [inline crates/bevy_dev_tools/Cargo.toml:3]:
```suggestion
version = "0.14.0-dev"
```

File: crates/bevy_dev_tools/src/lib.rs
Comment (BD103) [inline crates/bevy_dev_tools/src/lib.rs:1]:
```suggestion
```

This can be removed. `missing_docs` is now warn-by-default.

https://github.com/bevyengine/bevy/blob/f5ab1040a50edfe6f93791c28014915f49ea1933/Cargo.toml#L47-L49

File: crates/bevy_internal/Cargo.toml
Comment (BD103) [inline crates/bevy_internal/Cargo.toml:188]:
```suggestion
bevy_dev_tools = { path = "../bevy_dev_tools/", optional = true, version = "0.14.0-dev" }
```

File: crates/bevy_render/Cargo.toml
Comment (BD103) [inline crates/bevy_render/Cargo.toml:57]:
```suggestion
bevy_dev_tools = { path = "../bevy_dev_tools", version = "0.14.0-dev", optional = true }
```

File: crates/bevy_time/Cargo.toml
Comment (BD103) [inline crates/bevy_time/Cargo.toml:26]:
```suggestion
bevy_dev_tools = { path = "../bevy_dev_tools", version = "0.14.0-dev", optional = true }
```

**Review by BD103 — CHANGES_REQUESTED** [review]
A few changes requested, mainly with the bump to 0.14.0-dev.

**Comment by alice-i-cecile** [comment]
> https://github.com/bevyengine/bevy/pull/12305 is a feature that fits the best in bevy_dev_tools. What needs to be done for this PR to be merged, to unblock the simplified backtraces one?

So, IMO:

1. Resolve merge conflicts.
2. Bump to 0.14.0-dev.
3. Add a tool or two.
4. Get the final sign-off from @cart.

If there's no movement on this PR in the next week or so I'll adopt it and get it through :)

**Comment by pablo-lua** [comment]
If we want more tools here, we can move Aabb gizmos here, or merge some tools that are not merged yet, like #11237, and move here. I think gizmos has a lot of tools hanging out there that can be moved.

**Comment by matiqo15** [comment]
I've rebased onto main

**Review by cart — APPROVED** [review]
I'm happy with where this landed and how it is communicated to users in the documentation. Still slightly worried about the "grab bag" nature of this. We should pay careful attention to what ends up in here and consider spinning specific features out into their own crates as they mature / expand.

**Pull request merged**
